### DISCUSSION TITLE
failing testCantEnterWhenRaffleIsCalculating

### DISCUSSION THREAD

**Discussion by SquilliamX** [Cyfrin/foundry-full-course-cu#564] [Q&A] [answered]
Hi I am up to 5:50:43 when patrick compiles forge test -- match-test testCantEnterWhenRaffleIsCalculating and it is indeed reverting but without data. Can anyone help me pinpoint this problem?

Thanks in advance!

error message-

forge test --match-test testCantEnterWhenRaffleIsCalculating -vvvv
[⠆] Compiling...
No files changed, compilation skipped

Running 1 test for test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: Call reverted as expected, but without data] testCantEnterWhenRaffleIsCalculating() (gas: 148824)
Logs:
  Creating subscription on ChainId:  31337
  Your sub Id is  1
  PLease update subscriptionId in HelperConfig.s.sol
  Funding subscription: 1
  Using vrfCoordinator: 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
  On ChainID: 31337
  Adding consumer contract: 0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1
  Using vrfCoordinator:  0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
  On ChainID: 31337

Traces:
  [41283480] RaffleTest::setUp() 
    ├─ [20596225] → new DeployRaffle@0xFEfC6BAF87cF3684058D62Da40Ff3A795946Ab06
    │   └─ ← 102563 bytes of code
    ├─ [20455815] DeployRaffle::run() 
    │   ├─ [4508081] → new HelperConfig@0x6d2eed85750d316088343D6d5e91ca59eb052768
    │   │   ├─ [0] VM::startBroadcast() 
    │   │   │   └─ ← ()
    │   │   ├─ [1206889] → new VRFCoordinatorV2Mock@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   └─ ← 5805 bytes of code
    │   │   ├─ [746241] → new LinkToken@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38], amount: 1000000000000000000000000 [1e24])
    │   │   │   └─ ← 3267 bytes of code
    │   │   ├─ [0] VM::stopBroadcast() 
    │   │   │   └─ ← ()
    │   │   └─ ← 11642 bytes of code
    │   ├─ [963] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← 10000000000000000 [1e16], 30, VRFCoordinatorV2Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, 0, 500000 [5e5], LinkToken: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]
    │   ├─ [4986910] → new CreateSubscription@0x8Bbbb5d294c78bd70dd88E6D63498dECa0C7BBCe
    │   │   └─ ← 24684 bytes of code
    │   ├─ [53045] CreateSubscription::createSubscription(VRFCoordinatorV2Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]) 
    │   │   ├─ [0] console::9710a9d0(00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000007a6900000000000000000000000000000000000000000000000000000000000000224372656174696e6720737562736372697074696f6e206f6e20436861696e49643a20000000000000000000000000000000000000000000000000000000000000) [staticcall]
    │   │   │   └─ ← ()
    │   │   ├─ [0] VM::startBroadcast() 
    │   │   │   └─ ← ()
    │   │   ├─ [46789] VRFCoordinatorV2Mock::createSubscription() 
    │   │   │   ├─ emit SubscriptionCreated(subId: 1, owner: DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
    │   │   │   └─ ← 1
    │   │   ├─ [0] VM::stopBroadcast() 
    │   │   │   └─ ← ()
    │   │   ├─ [0] console::9710a9d0(00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000f596f757220737562204964206973200000000000000000000000000000000000) [staticcall]
    │   │   │   └─ ← ()
    │   │   ├─ [0] console::log(PLease update subscriptionId in HelperConfig.s.sol) [staticcall]
    │   │   │   └─ ← ()
    │   │   └─ ← 1
    │   ├─ [5084645] → new FundSubscription@0xffdcC297C11a15F778e155c20c0b7Ab1D26f5561
    │   │   └─ ← 25172 bytes of code
    │   ├─ [6970] FundSubscription::fundSubscription(VRFCoordinatorV2Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], 1, LinkToken: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]) 
    │   │   ├─ [0] console::9710a9d0(00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000001546756e64696e6720737562736372697074696f6e3a0000000000000000000000) [staticcall]
    │   │   │   └─ ← ()
    │   │   ├─ [0] console::log(Using vrfCoordinator:, VRFCoordinatorV2Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]) [staticcall]
    │   │   │   └─ ← ()
    │   │   ├─ [0] console::9710a9d0(00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000007a69000000000000000000000000000000000000000000000000000000000000000b4f6e20436861696e49443a000000000000000000000000000000000000000000) [staticcall]
    │   │   │   └─ ← ()
    │   │   ├─ [0] VM::startBroadcast() 
    │   │   │   └─ ← ()
    │   │   ├─ [3178] VRFCoordinatorV2Mock::fundSubscription(1, 3000000000000000000 [3e18]) 
    │   │   │   ├─ emit SubscriptionFunded(subId: 1, oldBalance: 0, newBalance: 3000000000000000000 [3e18])
    │   │   │   └─ ← ()
    │   │   ├─ [0] VM::stopBroadcast() 
    │   │   │   └─ ← ()
    │   │   └─ ← ()
    │   ├─ [0] VM::startBroadcast() 
    │   │   └─ ← ()
    │   ├─ [562582] → new Raffle@0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1
    │   │   └─ ← 2685 bytes of code
    │   ├─ [0] VM::stopBroadcast() 
    │   │   └─ ← ()
    │   ├─ [5024764] → new AddConsumer@0x7C5f18184dbdb4fE367fC3B38AaA576C6B86f2Af
    │   │   └─ ← 24873 bytes of code
    │   ├─ [50979] AddConsumer::addConsumer(Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1], VRFCoordinatorV2Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], 1) 
    │   │   ├─ [0] console::log(Adding consumer contract:, Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1]) [staticcall]
    │   │   │   └─ ← ()
    │   │   ├─ [0] console::log(Using vrfCoordinator: , VRFCoordinatorV2Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]) [staticcall]
    │   │   │   └─ ← ()
    │   │   ├─ [0] console::9710a9d0(00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000007a69000000000000000000000000000000000000000000000000000000000000000b4f6e20436861696e49443a000000000000000000000000000000000000000000) [staticcall]
    │   │   │   └─ ← ()
    │   │   ├─ [0] VM::startBroadcast() 
    │   │   │   └─ ← ()
    │   │   ├─ [47262] VRFCoordinatorV2Mock::addConsumer(1, Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1]) 
    │   │   │   ├─ emit ConsumerAdded(subId: 1, consumer: Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1])
    │   │   │   └─ ← ()
    │   │   ├─ [0] VM::stopBroadcast() 
    │   │   │   └─ ← ()
    │   │   └─ ← ()
    │   └─ ← 0x00000000000000000000000062c20aa1e0272312bc100b4e23b4dc1ed96dd7d10000000000000000000000006d2eed85750d316088343d6d5e91ca59eb052768
    ├─ [963] HelperConfig::activeNetworkConfig() [staticcall]
    │   └─ ← 10000000000000000 [1e16], 30, VRFCoordinatorV2Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, 0, 500000 [5e5], LinkToken: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]
    ├─ [0] VM::deal(player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C], 10000000000000000000 [1e19]) 
    │   └─ ← ()
    └─ ← ()

  [148824] RaffleTest::testCantEnterWhenRaffleIsCalculating() 
    ├─ [0] VM::prank(player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C]) 
    │   └─ ← ()
    ├─ [47734] Raffle::enterRaffle{value: 10000000000000000}() 
    │   ├─ emit EnteredRaffle(player: player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C])
    │   └─ ← ()
    ├─ [0] VM::warp(32) 
    │   └─ ← ()
    ├─ [0] VM::roll(2) 
    │   └─ ← ()
    ├─ [69944] Raffle::performUpkeep(0x) 
    │   ├─ [43604] VRFCoordinatorV2Mock::requestRandomWords(0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, 1, 3, 500000 [5e5], 1) 
    │   │   ├─ emit RandomWordsRequested(keyHash: 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, requestId: 1, preSeed: 100, subId: 1, minimumRequestConfirmations: 3, callbackGasLimit: 500000 [5e5], numWords: 1, sender: Raffle: [0x62c20Aa1e0272312BC100b4e23B4DC1Ed96dD7D1])
    │   │   └─ ← 1
    │   └─ ← ()
    ├─ [0] VM::expectRevert(Raffle__RaffleNotOpen()) 
    │   └─ ← ()
    ├─ [0] VM::prank(player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C]) 
    │   └─ ← ()
    ├─ [313] Raffle::enterRaffle{value: 10000000000000000}() 
    │   └─ ← "EvmError: Revert"
    └─ ← "Call reverted as expected, but without data"

Test result: FAILED. 0 passed; 1 failed; finished in 16.65ms

Failing tests:
Encountered 1 failing test in test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: Call reverted as expected, but without data] testCantEnterWhenRaffleIsCalculating() (gas: 148824)

Encountered a total of 1 failing tests, 0 tests succeeded
function testCantEnterWhenRaffleIsCalculating

 function testCantEnterWhenRaffleIsCalculating() public {
        vm.prank(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
        vm.warp(block.timestamp + interval + 1);
        vm.roll(block.number + 1);
        raffle.performUpkeep("");

        vm.expectRevert(Raffle.Raffle__RaffleNotOpen.selector);
        vm.prank(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
    }
}
DeployRaffle.s.sol-

// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {Raffle} from "../src/Raffle.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {CreateSubscription, FundSubscription, AddConsumer} from "./Interactions.s.sol";

contract DeployRaffle is Script {
    function run() external returns (Raffle, HelperConfig) {
        HelperConfig helperConfig = new HelperConfig();
        (
            uint256 entranceFee,
            uint256 interval,
            address vrfCoordinator,
            bytes32 gasLane,
            uint64 subscriptionId,
            uint32 callbackGasLimit,
            address link
        ) = helperConfig.activeNetworkConfig();

        if (subscriptionId == 0) {
            // we are going to need to create a subscription!
            CreateSubscription createSubscription = new CreateSubscription();
            subscriptionId = createSubscription.createSubscription(
                vrfCoordinator
            );

            // Fund it!
            FundSubscription fundSubscription = new FundSubscription();
            fundSubscription.fundSubscription(
                vrfCoordinator,
                subscriptionId,
                link
            );
        }

        vm.startBroadcast();
        Raffle raffle = new Raffle(
            entranceFee,
            interval,
            vrfCoordinator,
            gasLane,
            subscriptionId,
            callbackGasLimit
        );
        vm.stopBroadcast();

        AddConsumer addConsumer = new AddConsumer();
        addConsumer.addConsumer(
            address(raffle),
            vrfCoordinator,
            subscriptionId
        );
        return (raffle, helperConfig);
    }
}

raffle.sol-

// Layout of Contract:
// version
// imports
// errors
// interfaces, libraries, contracts
// Type declarations
// State variables
// Events
// Modifiers
// Functions

// Layout of Functions:
// constructor
// receive function (if exists)
// fallback function (if exists)
// external
// public
// internal
// private
// view & pure functions

// SPDX-License Identifier: MIT

pragma solidity ^0.8.18;

import {VRFCoordinatorV2Interface} from "@chainlink/contracts/src/v0.8/interfaces/VRFCoordinatorV2Interface.sol";
import {VRFConsumerBaseV2} from "@chainlink/contracts/src/v0.8/VRFConsumerBaseV2.sol";

/**
 * @title A sample Raffle contract
 * @author SmartChainwizz
 * @notice This is a sample contract for a raffle
 * @dev Implements Chainlink VRFv2
 */

contract Raffle is VRFConsumerBaseV2 {
    error Raffle__NotEnoughEthSent();
    error Raffle__TransferFailed();
    error Raffle__RaffleNotOpen();
    error Raffle__UpkeepNotNeeded(
        uint256 currentBalance,
        uint256 numPlayers,
        uint256 raffleState
    );

    // bool lotteryState = open, closed, calcualating
    /* Type Declarations */
    enum RaffleState {
        OPEN, // 0
        CALCULATING // 1
        // 2 <-- CAN ADD MORE OPTIONS TO ENUM IF NEEDED
        // 3
    }

    //* State Variables */
    uint16 private constant REQUEST_CONFIRMATIONS = 3;
    uint32 private constant NUM_WORDS = 1;

    uint256 private immutable i_entranceFee;
    // @dev Duration of the lottery in seconds
    uint256 private immutable i_interval;
    VRFCoordinatorV2Interface private immutable i_vrfCoordinator;
    bytes32 private immutable i_gasLane;
    uint64 private immutable i_subscriptionId;
    uint32 private immutable i_callbackGasLimit;

    address payable[] private s_players;
    uint256 private s_lastTimeStamp;
    address private s_recentWinner;
    RaffleState private s_raffleState;

    /** Events */
    event EnteredRaffle(address indexed player);
    event PickedWinner(address indexed winner);

    constructor(
        uint256 entranceFee,
        uint256 interval,
        address vrfCoordinator,
        bytes32 gasLane,
        uint64 subscriptionId,
        uint32 callbackGasLimit
    ) VRFConsumerBaseV2(vrfCoordinator) {
        i_entranceFee = entranceFee;
        i_interval = interval;
        i_vrfCoordinator = VRFCoordinatorV2Interface(vrfCoordinator);
        i_gasLane = gasLane;
        i_subscriptionId = subscriptionId;
        i_callbackGasLimit = callbackGasLimit;

        s_raffleState = RaffleState.OPEN;
        s_lastTimeStamp = block.timestamp;
    }

    function enterRaffle() external payable {
        //require(msg.value >= i_entranceFee, "Not enough ETH sent");
        if (msg.value < i_entranceFee) {
            revert Raffle__NotEnoughEthSent();
        }

        if (s_raffleState != RaffleState.OPEN) {
            revert();
        }

        s_players.push(payable(msg.sender));
        emit EnteredRaffle(msg.sender);
    }

    // When is the winner supposed to be picked?
    /**
     * @dev This is the function that the ChainLink Automation nodes call to see if its time to preform an upkeep.
     * the fillowing should be true for this to return true:
     * 1. The time interval has passed between raffel runs
     * 2. The raffle is in an open state
     * 3. the contract has ETH(aka, players)
     * 4. (Implicit) the subcription is funded with Link
     */
    function checkUpkeep(
        bytes memory /* checkData */
    ) public view returns (bool upkeepNeeded, bytes memory /* performData */) {
        bool timeHasPassed = ((block.timestamp - s_lastTimeStamp) >=
            i_interval);
        bool isOpen = RaffleState.OPEN == s_raffleState;
        bool hasbalance = address(this).balance > 0;
        bool hasplayers = s_players.length > 0;
        upkeepNeeded = (timeHasPassed && isOpen && hasbalance && hasplayers);
        return (upkeepNeeded, "0x0");
    }

    //1. Get a random number
    //2. use the random number to pick a player
    //3. be automatically called
    function performUpkeep(bytes calldata /* performData */) external {
        (bool upkeepNeeded, ) = checkUpkeep("");
        if (!upkeepNeeded) {
            revert Raffle__UpkeepNotNeeded(
                address(this).balance,
                s_players.length,
                uint256(s_raffleState)
            );
        }
        //check to see if enough time has passed

        // 1000 - 500 = 500. 600 seconds Not enough time has passed (Example)
        // 1200 - 500 = 700. 600 seconds. Passed the 600seconds interval.(Example)

        s_raffleState = RaffleState.CALCULATING;

        // Chainlink VRF is 2 transactions. -
        // 1. Request a random number from the Chainlink VRF
        // 2. Get the random number from the Chainlink VRF

        i_vrfCoordinator.requestRandomWords(
            i_gasLane, // gas lane
            i_subscriptionId, //
            REQUEST_CONFIRMATIONS,
            i_callbackGasLimit,
            NUM_WORDS
        );
    }

    // CEI: Checks, Effects, Interactions
    function fulfillRandomWords(
        uint256 /* requestId*/,
        uint256[] memory randomWords
    ) internal override {
        // checks (require statements, ifs, errors)
        // effects (our own contract)
        uint256 indexOfWinner = randomWords[0] % s_players.length;
        address payable winner = s_players[indexOfWinner];
        s_recentWinner = winner;
        s_raffleState = RaffleState.OPEN;

        s_players = new address payable[](0);
        s_lastTimeStamp = block.timestamp;
        // Interactions (other contracts)
        (bool success, ) = winner.call{value: address(this).balance}("");
        if (!success) {
            revert Raffle__TransferFailed();
        }
        emit PickedWinner(winner);
    }

    /** Getter Functions */

    function getEntranceFee() external view returns (uint256) {
        return i_entranceFee;
    }

    function getRaffleState() external view returns (RaffleState) {
        return s_raffleState;
    }

    function getPlayer(uint256 indexOfPlayer) external view returns (address) {
        return s_players[indexOfPlayer];
    }
}

**Comment by Niferu** [comment]
@smartchainwiz 

Have you funded your `PLAYER` using `deal()` function?

Please also share your repository link, it will be much easier to find error.

**Comment by theirrationalone** [comment, marked as answer]
I think you created two identical discussions... 

Try this:

[Try this](https://github.com/Cyfrin/foundry-full-course-f23/discussions/563#discussioncomment-6654213)

**Comment by SquilliamX** [comment]
thanks this was the error!
